re-enable Istio test, other misc. code fixes around endpoint

ianvernon opened the pull request:
Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6467)
<!-- Reviewable:end -->

Commit:
endpoint: misc. logging-related fixes and enhancements

Update some logging statements during the lifecycle of the endpoint to make it
easier to correlate when certain events occur for an endpoint in the
cilium-agent logs. Along with this, update the endpoint logger with the numeric
identity of the endpoint when the endpoint's identity is changed, and also
update the logger when the endpoint's node-local ID is updated. This fixed an
issue where some logs for an endpoint would show it having endpoint ID 0 even
after the endpoint was allocated an ID.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: re-enable Istio test

Also remove init policy, as there is no longer an init-identity phase for
endpoints in Cilium when Cilium is used in tandem with Kubernetes.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test: add wait for endpoints to be ready in policy enforcement

Endpoints may not be ready after the container is created, and
as such their policy enforcement fields may not be updated yet
as well. To reduce the possibility of flakes in the CI, wait
for all endpoints to be ready before checking the enforcement
status for ingress and egress of running endpoints in the
policy enforcement Ginkgo test.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -422,22 +422,21 @@ func (e *Endpoint) Regenerate(owner Owner, regenMetadata *ExternalRegenerationMe
 		doneFunc := owner.QueueEndpointBuild(uint64(e.ID))
 		if doneFunc != nil {
-			scopedLog.Debug("Dequeued endpoint from build queue")
+			e.getLogger().Debug("Dequeued endpoint from build queue")
```
Please leave it as it was before, there's no need to call atomic.Load 3 times.

tgraf replied:
I think at least before and after `regenerate()`, the scopedLog must be refreshed as it will have changed significantly. We should optimize for correctness in the logs instead of micro-optimizing this code at this level. If we want to optimize then let's ensure that we don't call `getLogger()` when debugging is not enabled.

jrajahalme replied:
Nit-picking here, but `getLogger()` is called on average about 1 time only. On monitorAPI failure it's 2 times, but there is no need to optimize that.

ianvernon replied:
The reason I did this is because there is a lot of change in state to the endpoint during this process, and I found that information about the endpoint was stale in the log (e.g., endpoint ID was zero at certain points in time even after it had been added to the endpoint manager and allocated an ID). I think the only time we can safely use a scoped log is if the endpoint is locked, and the endpoint structure isn't locked here.

aanm commented on daemon/endpoint.go:
```diff
@@ -231,7 +231,7 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 	}
 
-	log.Debug("Synchronously waiting for endpoint to regenerate")
+	ep.Logger(daemonSubsys).Debug("Synchronously waiting for endpoint to regenerate")
```
These changes seem out of scope of commit message

ianvernon replied:
I'll update the commit message explaining what I've done here.

aanm commented on daemon/daemon_main.go:
```diff
@@ -87,11 +87,12 @@ const (
 
 var (
-	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "daemon")
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, daemonSubsys)
```
These changes seem out of scope of commit message

ianvernon replied:
I'll update the commit message explaining what I've done here.

ianvernon replied:
@aanm updated commit message.

aanm commented on pkg/endpointmanager/manager.go:
```diff
@@ -79,8 +79,14 @@ func Insert(ep *endpoint.Endpoint) error {
+		// Update endpoint logger to contain ID.
+		ep.UnconditionalRLock()
+		ep.UpdateLogger(map[string]interface{}{
+			logfields.EndpointID: ep.ID,
```
ep.ID doesn't need to be locked

ianvernon replied:
Ah, did not realize! I will add a comment to the ID field in the Endpoint to note this. Will update accordingly.

ianvernon replied:
Hm, but `UpdateLogger` documentation says to hold the endpoint's mutex for reading? Should that statement be removed from its documentation?

aanm replied:
@ianvernon ah, I see. The comment should say: `if fields is nil you must hold Endpoint.Mutex for reading.`

ianvernon replied:
Ah, got it! Will remove the `RLock` and `RUnlock` in next upload.

aanm commented on pkg/endpointmanager/manager.go:
```diff
@@ -399,10 +405,12 @@ func AddEndpoint(owner endpoint.Owner, ep *endpoint.Endpoint, reason string) (er
-			return err
-		}
+		go func() {
+			if err := ep.RegenerateWait(owner, reason); err != nil {
```
If this is inside a go routine how will we return to the caller the regeneration has failed?

tgraf replied:
Callers could not rely on this as it was dependent on the state already. We only have two callers to `AddEndpoint()`:

1. `createEndpoint()`: When new endpoints are inserted
2. `LaunchAsEndpoint()` when the health endpoint is created.

We should move to `RegenerateWait()` into the callers:

* For `LaunchAsEndpoint()` it should run as a controller to retry on error and also have an interval for rebuilds such as 15 minutes so we have a fallback

* For `createEndpoint()` it should also start a controller to keep retrying to build the endpoint. We can do an interval of 15 minutes as well. When `createEndpoint()` is in `SyncBuildEndpoint` mode then it should first do the sync build and then start the controller.

The controller must be removed on endpoint leave.

Once we have that we can follow-up and use the controller for all endpoint regenerations and we always have retries on build failures.

jrajahalme replied:
+1 for the retries on build failures!

ianvernon replied:
I wasn't sure whether that was in the scope of this PR or not. I wanted to minimize the existing changes in behavior in the cilium-agent during one of the most critical parts of code (creation of endpoints) here, so I made the regeneration of the endpoint asynchronous, but still removed it upon failure if the first build failed, which preserves existing behavior as much as possible. I did this because this needs to be backported to the v1.4 branch. @tgraf do you want this change to be made as part of this PR? I'm worried that we may change behavior in a way we don't expect with that sort of a change, which may be better to do as part of the work for v1.5. I may be overly cautious here, though.

aanm commented on pkg/endpointmanager/manager.go:
```diff
@@ -405,12 +405,20 @@ func AddEndpoint(owner endpoint.Owner, ep *endpoint.Endpoint, reason string) (er
 
 	if build {
-		go func() {
+		if asynchronouslyBuild {
```
I don't see a good reason to add the logic of building the endpoint async or not into the endpoint manager. Besides, the endpoint itself offers a function to perform the regeneration async, `ep.Regenerate`.

ianvernon replied:
I did this to maintain existing behavior on the Docker plugin side, so I would be reducing the amount of changes introduced in this PR. I also kept the `RegenerateWait` call in so we would remove the endpoint in the async case if the first build failed, as was being done before since I didn't want to change existing behavior too much in that case as well.

aanm commented on pkg/endpointmanager/manager.go:
```diff
@@ -405,12 +405,20 @@ func AddEndpoint(owner endpoint.Owner, ep *endpoint.Endpoint, reason string) (er
+			go func() {
+				if err := ep.RegenerateWait(owner, reason); err != nil {
+					ep.Logger("endpointmanager").Error("removing endpoint because initial build failed")
+					Remove(ep)
```
It seems odd to me to regenerate an endpoint and if something fails we remove it from the endpointmanager without having the caller of this function to be aware of that.

ianvernon replied:
Yeah, I was a bit unsure about what to do here, so I erred on the side of caution by keeping the existing behavior of removing the endpoint from the EndpointManager if the first build for it failed :/

tgraf commented on pkg/endpointmanager/manager.go:
```diff
@@ -405,12 +405,20 @@ func AddEndpoint(owner endpoint.Owner, ep *endpoint.Endpoint, reason string) (er
 
 	if build {
-		go func() {
+		if asynchronouslyBuild {
```
So it seems that for the docker plugin, we have always been waiting for the initial build to complete. It would be better then for the docker plugin to also pass in `SyncBuildEndpoint`. The docker plugin version is tied to the agent version 1:1 so as we change the behavior of `createEndpoint()`, it is fine to require changing the docker plugin as well. In particular as an API user would expect an async build if `SyncBuildEndpoint` is *not* passed in.

tgraf commented on test/runtime/Policies.go:
```diff
@@ -158,6 +158,8 @@ var _ = Describe("RuntimePolicyEnforcement", func() {
 
 			By("Create a new container")
 			vm.ContainerCreate("new", helpers.HttpdImage, helpers.CiliumDockerNetwork, "-l id.new")
+			areEndpointsReady := vm.WaitEndpointsReady()
```
Please add a statement to the commit message why this is required.

ianvernon replied:
Fixed in latest upload.

tgraf reviewed (changes requested):
Fantastic work. Really glad to see that we got it working without init policies.

jrajahalme reviewed (approved):
Great work! Nothing to add to the thorough reviews by @aanm and @tgraf.

tgraf committed:
endpoint: document why async regeneration is required on first build

If the entire regeneration for an endpoint is waited to be completed (e.g.,
policy generation, BPF compilation / loading, and proxy updates) when an
endpoint is created, then creation of endpoints which have L7 policy which
selects them when Cilium is running with Istio will result in a deadlock. This
is because when a pod is created, its sidecar proxy is not started yet, so when
addition of the endpoint into the endpoint manager waits for regeneration to
complete, the regeneration gets stuck because the its trying to configure a
proxy which does not exist it. To get around this, do not synchronously wait
for the entire regeneration process to succeed for a given endpoint in
endpointmanager. Add a comment here so this is not changed in the future / to
provide context to future editors of this code.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on daemon/endpoint.go:
```diff
@@ -211,19 +211,37 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
+		return PutEndpointIDFailedCode, fmt.Errorf("endpoint was deleted while waiting for the identity")
+	}
+
+	build := ep.GetStateLocked() == endpoint.StateReady
```
I don't understand this line, why is the endpoint only being built if it's ready? AFAIK the CNI plugin sends the endpoint to the cilium-agent in the `endpoint.StateWaitingToRegenerate` state

ianvernon replied:
cc: @tgraf  - I cherry-picked this commit per his recommendation. @tgraf can you provide some context here? I can add it as a comment for future reference.

ianvernon replied:
This change is no longer part of this PR so marking as resolved.

aanm commented on daemon/endpoint.go:
```diff
@@ -238,10 +238,21 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
-			return PutEndpointIDFailedCode, fmt.Errorf("failed to build the endpoint: %s", err)
-		}
+		// Do not synchronously regenerate the endpoint when adding it into the
+		// endpointmanager. We have custom logic later for waiting for specific
```
nit, this is not endpointmanager.

ianvernon replied:
Will be fixed in next upload.

Commit:
endpoint: document when mutex not needed for updating logger

The endpoint mutex does not need to be acquired if the fields parameter is non-nil when calling
`endpoint.UpdateLogger()`, so document it accordingly.

Signed-off by: Ian Vernon <ian@cilium.io>

tgraf commented on cilium-health/launch/endpoint.go:
```diff
@@ -266,10 +266,15 @@ func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressin
 		return fmt.Errorf("Error while adding endpoint: %s", err)
 	}
 
+	if err := ep.LockAlive(); err != nil {
```
Please add a `Fixes: ` tag to the commit message

ianvernon replied:
Done and posted as a separate PR. I'd still like this backported to v1.4 because I don't want to break the Istio integration again.

tgraf reviewed (changes requested):
Please separate out the fixes, in particular, the last commit fixing up the missing lock in the health daemon so we can backport. I doubt that we want to backport all commits of this PR into all branches.
